content/reference/programming-model.md

```diff
@@ -962,9 +962,93 @@ We can now create instances of the new `MyResource` resource kind in our program
 
+### Dynamic Resource Inputs
+
+The interface to other components outside of the dynamic provider is the `pulumi.dynamic.Resource`. To pass inputs to your implementation of the `pulumi.dynamic.ResourceProvider`, you actually need to pass them through the resource that is exposed to the outside world. That is the `props` parameter in the constructor in the example below. Any properties you set in the `props` object will be passed to your `pulumi.dynamic.ResourceProvider` functions as the `inputs` where appropriate.
+
+> For statically typed languages, you can get rich type information for the inputs declaring an input type to make it easy to use your custom resource.
+
+{{< langchoose >}}
+
+```javascript
+class MyResource extends pulumi.dynamic.Resource {
+    constructor(name, props, opts) {
+        super(myprovider, name, props, opts); 
+    }
+}
+```
+
+```typescript
+interface MyResourceInputs {
+    myStringProp: pulumi.Input<string>;
+    myBoolProp: pulumi.Input<boolean>;
+    ...
+}
+
+class MyResource extends pulumi.dynamic.Resource {
+    constructor(name: string, props: MyResourceInputs, opts?: pulumi.CustomResourceOptions) {
+        super(myprovider, name, props, opts); 
+    }
+}
+```
+
+```python
+# Dynamic Providers are not supported in Python currently.
+```
+
+```go
+// Dynamic Providers are not supported in Go currently.
+```
+
 ### Resource Provider Interface
 
-Implementing the `pulumi.dynamic.ResourceProvider` interface requires implementing a subset of the methods below.  Each of these methods is asynchronous, and most common implementations of these methods will do asynchronous network I/O to provision resources in a backing cloud provider or other resource model. There are several important contracts to be aware of as part of determining how the Pulumi engine will call each of these methods, and with what data.
+Implementing the `pulumi.dynamic.ResourceProvider` interface requires implementing a subset of the methods below. Each of these methods is asynchronous, and most common implementations of these methods will do asynchronous network I/O to provision resources in a backing cloud provider or other resource model. There are several important contracts to be aware of as part of determining how the Pulumi engine will call each of these methods, and with what data.
+
+Since the inputs to the custom resource are of type `pulumi.Input<T>`, when the functions below are called with those inputs they will be passed the resolved version of the same inputs. To get strong typing for your inputs in the provider functions, you should have another type that contains the same properties but with unwrapped-types as their types.
```

Comment:
```suggestion
Though the input properties passed to a `pulumi.dynamic.Resource` instance will usually be of type `pulumi.Input<T>` in order to compose well with other Pulumi resources, the dynamic provider's functions are invoked with the fully resolved input values. Strong typing for the inputs to your provider's functions can help clarify this; you can achieve this by creating a second interface with the same properties as your resource's inputs, but with fully unwrapped types.
```

Reply:
I love this suggestion. Makes it read much better I think. Thanks!